src/hashtable.c

```diff
@@ -0,0 +1,2165 @@
+        }
+
+        /* Iterate over indices in larger table that are the expansion of the
+         * index pointed to by the cursor in the smaller table. */
+        do {
+            /* Emit entries in the larger table at this cursor, if this index
+             * hash't already been rehashed. */
+            idx = cursor & mask_large;
+            if (table_large == 1 || ht->rehash_idx == -1 || idx >= (size_t)ht->rehash_idx) {
+                size_t used_before = ht->used[table_large];
+                bucket *b = &ht->tables[table_large][idx];
+                do {
+                    if (b->presence) {
+                        for (int pos = 0; pos < ENTRIES_PER_BUCKET; pos++) {
+                            if (isPositionFilled(b, pos)) {
+                                void *emit = emit_ref ? &b->entries[pos] : b->entries[pos];
+                                fn(privdata, emit);
+                            }
+                        }
+                    }
+                    bucket *next = bucketNext(b);
+                    if (next != NULL && defragfn != NULL) {
+                        next = bucketDefrag(b, next, defragfn);
+                    }
+                    b = next;
+                } while (b != NULL);
+                /* If any entries were deleted, fill the holes. */
+                if (ht->used[table_large] < used_before) {
+                    compactBucketChain(ht, idx, table_large);
+                }
+            }
+
+            /* Increment the reverse cursor not covered by the smaller mask. */
+            cursor = nextCursor(cursor, mask_large);
+
+            /* Continue while bits covered by mask difference is non-zero. */
+        } while (cursor & (mask_small ^ mask_large));
+    }
+    hashtableResumeRehashing(ht);
+    return cursor;
+}
+
+/* --- Iterator --- */
+
+/* Initiaize a iterator, that is not allowed to insert, delete or even lookup
+ * entries in the hashtable, because such operations can trigger incremental
+ * rehashing which moves entries around and confuses the iterator. Only
+ * hashtableNext is allowed. Each entry is returned exactly once. Call
+ * hashtableResetIterator when you are done. See also
+ * hashtableInitSafeIterator. */
+void hashtableInitIterator(hashtableIterator *iterator, hashtable *ht) {
+    iter *iter;
+    iter = iteratorFromOpaque(iterator);
+    iter->hashtable = ht;
+    iter->table = 0;
+    iter->index = -1;
+    iter->safe = 0;
+}
+
+/* Initialize a safe iterator, which is allowed to modify the hash table while
+ * iterating. It pauses incremental rehashing to prevent entries from moving
+ * around. Call hashtableNext to fetch each entry. You must call
+ * hashtableResetIterator when you are done with a safe iterator.
+ *
+ * It's allowed to insert and replace entries. Deleting entries is only allowed
+ * for the entry that was just returned by hashtableNext. Deleting other entries
+ * are not supported. (It can cause internal fragmentation.)
```

Comment:
Not sure what you mean. The code that handles internal fragmentation in iterator and scan only works for the bucket chain that was just iterated/scanned. If the user deletes from another bucket chain while iterating or from the scan callback, it will leave holes that we don't handle.